cli/vscodeipc/vscodeipc.go

```diff
@@ -0,0 +1,268 @@
+package vscodeipc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/google/uuid"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/xerrors"
+	"tailscale.com/tailcfg"
+
+	"github.com/coder/coder/agent"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/codersdk"
+)
+
+// New creates a VS Code IPC client that can be used to communicate with workspaces.
+//
+// Creating this IPC was required instead of using SSH, because we're unable to get
+// connection information to display in the bottom-bar when using SSH. It's possible
+// we could jank around this (maybe by using a temporary SSH host), but that's not
+// ideal.
+//
+// This persists a single workspace connection, and lets you execute commands, check
+// for network information, and forward ports.
+func New(ctx context.Context, client *codersdk.Client, agentID uuid.UUID, options *codersdk.DialWorkspaceAgentOptions) (http.Handler, io.Closer, error) {
+	if options == nil {
+		options = &codersdk.DialWorkspaceAgentOptions{}
+	}
+	// We need this to track upload and download!
+	options.EnableTrafficStats = true
+
+	agentConn, err := client.DialWorkspaceAgent(ctx, agentID, options)
+	if err != nil {
+		return nil, nil, err
+	}
+	api := &api{
+		agentConn: agentConn,
+	}
+	r := chi.NewRouter()
+	r.Get("/port/{port}", api.port)
+	r.Get("/network", api.network)
+	r.Post("/execute", api.execute)
+	return r, api, nil
+}
+
+type api struct {
+	agentConn     *codersdk.AgentConn
+	sshClient     *ssh.Client
+	sshClientErr  error
+	sshClientOnce sync.Once
+
+	lastNetwork time.Time
+}
+
+func (api *api) Close() error {
+	if api.sshClient != nil {
+		api.sshClient.Close()
+	}
+	return api.agentConn.Close()
+}
+
+type NetworkResponse struct {
```

Comment:
Is this API documented somewhere or should we document it (ping/network/execute)?

Reply:
It's used by the VS Code extension, but I'll add some comments on what it means to change it, and how to change it properly.